File: firmware/scripts/check_hardcoded_versions.rb
def read_software_common_versions
  text = File.read('../software/src/defs/Versions.ts')

  config_storage_format_revision = text.match(/^export const ConfigStorageFormatRevision = (\d+);$/)[1].to_i
  rawhid_message_protocol_revision = text.match(/^export const RawHidMessageProtocolRevision = (\d+);$/)[1].to_i

  {
    config_storage_format_revision: config_storage_format_revision,
    rawhid_message_protocol_revision: rawhid_message_protocol_revision
  }
end

def read_firmware_common_versions
  text = File.read('src/modules/versions.h')

  config_storage_format_revision = text.match(/^#define CONFIG_STORAGE_FORMAT_REVISION (\d+)$/)[1].to_i
  rawhid_message_protocol_revision = text.match(/^#define RAWHID_MESSAGE_PROTOCOL_REVISION (\d+)$/)[1].to_i
  {
    config_storage_format_revision: config_storage_format_revision,
    rawhid_message_protocol_revision: rawhid_message_protocol_revision
  }
end

if __FILE__ == $PROGRAM_NAME
  puts 'check version defintions ...'
  v_soft = read_software_common_versions
  v_firm = read_firmware_common_versions
  valid = v_soft == v_firm
  unless valid
    warn 'incompatibe version definitions'
    puts 'software:', v_soft
    puts 'firmware:', v_firm
    exit(1)
  end
  print "\e[A\e[K"
  puts 'check version defintions ... ok'
end

File: firmware/scripts2/checkHardcodedVersions.ts
import {
  checkHasFields,
  compareObjectByJsonStringifyParse,
  fsxReadTextFile,
  getMatched,
} from "./helpers";

process.chdir("..");

const sourceFilePaths = {
  softwareVersions: "../software/src/defs/Versions.ts",
  firmwareVersions: "./src/modules/versions.h",
};

interface ISynchronousVersionsSet {
  configStorageFormatRevision: string;
  rawHidProtocolRevision: string;
}

const requiredFields: (keyof ISynchronousVersionsSet)[] = [
  "configStorageFormatRevision",
  "rawHidProtocolRevision",
];

function readSoftwareCommonVersions(): ISynchronousVersionsSet {
  const text = fsxReadTextFile(sourceFilePaths.softwareVersions);
  return {
    configStorageFormatRevision: getMatched(
      text,
      /^export const ConfigStorageFormatRevision = (\d+);$/m
    )!,
    rawHidProtocolRevision: getMatched(
      text,
      /^export const RawHidMessageProtocolRevision = (\d+);$/m
    )!,
  };
}

function readFirmwareCommonVersions(): ISynchronousVersionsSet {
  const text = fsxReadTextFile(sourceFilePaths.firmwareVersions);
  return {
    configStorageFormatRevision: getMatched(
      text,
      /^#define CONFIG_STORAGE_FORMAT_REVISION (\d+)$/m
    )!,
    rawHidProtocolRevision: getMatched(
      text,
      /^#define RAWHID_MESSAGE_PROTOCOL_REVISION (\d+)$/m
    )!,
  };
}

function checkVersions() {
  console.log("check version defintions ...");
  const softwareVersions = readSoftwareCommonVersions();
  const firmwareVersions = readFirmwareCommonVersions();
  const vSoftOk = checkHasFields(softwareVersions, requiredFields);
  const vFirmOk = checkHasFields(firmwareVersions, requiredFields);
  const equivalent = compareObjectByJsonStringifyParse(
    softwareVersions,
    firmwareVersions
  );
  const valid = vSoftOk && vFirmOk && equivalent;
  if (!valid) {
    console.error("incompatibe version definitions");
    console.error({ softwareVersions, firmwareVersions });
    process.exit(1);
  }
  console.log("check version defintions ... ok");
}

checkVersions();
